extensions-contrib/kafka-emitter/src/main/java/org/apache/druid/emitter/kafka/KafkaEmitterConfig.java

```diff
@@ -47,19 +65,72 @@
-    this.alertTopic = Preconditions.checkNotNull(alertTopic, "alert.topic can not be null");
-    this.requestTopic = requestTopic;
+    validateEventTypes(eventTypes);
+    validateSegmentMetadataTopicFormat(segmentMetadataTopicFormat);
+    // Unless explicitly overridden, kafka emitter will always emit metrics and alerts
+    this.eventTypes = eventTypes == null ? new HashSet<>(Arrays.asList(METRICS, ALERTS)) : eventTypes;
+
+    // If the topics are configured, but entry is missing in event.type, then add that event to eventTypes
+    checkForMissingEventType(metricTopic, METRICS);
+    checkForMissingEventType(alertTopic, ALERTS);
+    checkForMissingEventType(requestTopic, REQUESTS);
+    checkForMissingEventType(segmentMetadataTopic, SEGMENT_METADATA);
+
+    this.metricTopic = this.eventTypes.contains(METRICS) ? Preconditions.checkNotNull(metricTopic, "metric.topic can not be null") : null;
+    this.alertTopic = this.eventTypes.contains(ALERTS) ? Preconditions.checkNotNull(alertTopic, "alert.topic can not be null") : null;
+    this.requestTopic = this.eventTypes.contains(REQUESTS) ? Preconditions.checkNotNull(requestTopic, "request.topic can not be null") : null;
+    this.segmentMetadataTopic = this.eventTypes.contains(SEGMENT_METADATA) ? Preconditions.checkNotNull(segmentMetadataTopic, "segment.metadata.topic can not be null") : null;
     this.clusterName = clusterName;
     this.kafkaProducerConfig = kafkaProducerConfig == null ? ImmutableMap.of() : kafkaProducerConfig;
+
+    // If segment.metadata.topic.format is not configured, use json as default like other event types
+    this.segmentMetadataTopicFormat = segmentMetadataTopicFormat == null ? JSON : segmentMetadataTopicFormat;
+    // Metadata in non-JSON format must go their own topic
+    if (this.segmentMetadataTopic != null && !JSON.equals(this.segmentMetadataTopicFormat)) {
+      this.segmentMetadataTopic = this.segmentMetadataTopic + "-" + this.segmentMetadataTopicFormat;
+    }
+  }
+
+  private void validateEventTypes(Set<String> eventTypes)
+  {
+    if (eventTypes != null) {
+      List<String> validEvents = Arrays.asList(METRICS, REQUESTS, ALERTS, SEGMENT_METADATA);
+      for (String event : eventTypes) {
+        if (!validEvents.contains(event)) {
+          throw new InvalidConfigurationException("event.types has an invalid value " + event);
+        }
+      }
+    }
+  }
+
+  private void validateSegmentMetadataTopicFormat(String segmentMetadataTopicFormat)
+  {
+    if (segmentMetadataTopicFormat != null) {
+      List<String> validFormats = Arrays.asList(JSON, PROTOBUF);
+      if (!validFormats.contains(segmentMetadataTopicFormat)) {
+        throw new InvalidConfigurationException("metadata.topic.format has an invalid value " + segmentMetadataTopicFormat);
+      }
+    }
+  }
+
+  // Incase topic name is set, then add it to event types
+  // This would help if the topic name is set, but the event type is not set correctly
```

Comment:
why? it's a bit odd to expect someone might incorrectly configure something. Instead we should make the configuration more obvious. I think there is a valid use-case to leave the topic name but to temporarily disable some event types. Otherwise we might as well remove the configuration and instead rely only on which topic names are set. However I do think it's nice to have them separate since this would allow a default set of topic names if we wanted to.

Reply:
ok, in that case what is the purpose of having the "eventType" config if we always add the event types for topics that are configured? Doesn't this mean that the topic configs alone define which events are being collected, and we can get rid of eventTypes?

Reply:
Good point! I'll explain it in detail here

#### Existing behaviour:
By default `metrics and alerts` are sent out and requests are sent out if `requests.topic` is set. So, `metric.topic` and `alerts.topic` are mandatory fields and cannot be left empty.


#### New behavior:
##### New configs added:
1) `eventTypes`: This has default value of `(metrics, alerts)` to not break backward compatibility. So, `metric.topic` and `alerts.topic` cannot be left empty, if we are using default value for `eventTypes`.
But now, if someone explicitly wants to override evenTypes (for example, in our case, we want kafka emitter to emit only metadata-related info, so, we would set eventType to `(segment_metadata)`.  So, here, its valid to leave `metrics.topic` and `alerts.topic` to be empty. The eventType field is used to differentiate between these two cases.
2) `segmentMetadataTopicFormat`: This decides the format in which segment metadata is published by the emitter.

Reply:
ok, I understand part of the motivation now, but this makes it really confusing to reason about those configs for someone that doesn't have the historical context and tries to configure a new deployment.

For instance, with the way this is implemented right now:
- If I set a topic for an event type, but omit it from eventTypes, then I still get events for that type

So removing an entry from eventTypes config alone is not enough to prevent sending that particular type of data. I also have to remove the topic config, which is not intuitive at all, and seems to defeat the purpose of having a config for eventTypes.

A good test is – how would we write the documentation for this eventType config, since on its own it isn't enough to really specify the types of events you want. If that confusing then we probably don't have the right abstraction.

In my opinion a more intuitive approach would be to preserve the existing behavior of including alert and metrics if no eventType config is specified. However once eventType is specified, we no longer have to preserve the old behavior – since that config didn't exist before – and only the specified eventTypes get sent, regardless of the topic settings. Topics only become optional if they are not defined as part of eventType, which is also consistent with saying that eventTypes defaults to (alert, metrics), but I can remove an eventType from the list without having to worry about what the topic settings say.